### PR TITLE
[TESTING] cache AWS credentials to avoid hitting rate limits on EC2 instances with role-based auth

### PR DESCRIPTION
**Pull request by boris-dimitrov**
Thanks Andrey Kislyuk and David Rissato Cruz

### CODE REVIEW

File: scripts/s3mi [hunk 1]
```diff
@@ -326,6 +330,34 @@ def adjust_RAM_params():
...
+    On EC2/ECS, the AWS CLI calls the instance/container metadata service to fetch instance profile/role credentials
+    every time it runs, unless it finds credentials in environment variables or config files.
+
+    There is a rate limit on the metadata service, and we trigger it sometimes with the following error:
+    Error when retrieving credentials from container-role: Error retrieving metadata: Received non 200 response
+    (429) from ECS metadata: You have reached maximum request limit.
+
+    To avoid this, we fetch the credentials using standard botocore logic and cache them in a process-wide cache.
+    These credentials are short-lived, so we refresh them if we last cached them more than 10 minutes ago.
+    This behavior is only active when the constructor is called with cache_aws_credentials=True.
```
Comment (kislyuk) [inline scripts/s3mi]: I don't think you're using `cache_aws_credentials`, so probably best to remove this line.

Comment (boris-dimitrov) [reply scripts/s3mi]: Thanks

File: scripts/s3mi [hunk 2]
```diff
@@ -25,6 +25,15 @@ except:
     DEVNULL = open(os.devnull, "r+b")
 
 
+# for credential caching
+try:
+    import botocore.session
+    have_boto = True
+except:
+    have_boto = False
```
Comment (boris-dimitrov) [inline scripts/s3mi:33]: @kislyuk Do you have any suggestions for this case?   Seems to affect python2 environments.  My current approach is to just not do the session credential caching in those envs.

**Comment by boris-dimitrov** [comment]
Completed testing on idseq staging environment in two ways:    1) via cloud init change for gsnap asg to use this branch of s3mi;   also 2) via idseq-dag custom branch to use this branch in sample https://staging.idseq.net/samples/13350/pipeline_runs

Both tests succeeded.

**Comment by boris-dimitrov** [comment]
The testing covers both python 2 (gsnap ASG) and python3 (dag batch).

File: scripts/s3mi [hunk 3]
```diff
@@ -326,14 +338,46 @@ def adjust_RAM_params():
         MAX_CONCURRENT_REQUESTS = 15
 
 
-def main_cat(s3_uri):
+def get_credentials():
```
Comment (boris-dimitrov) [inline scripts/s3mi:341]: Question for @kislyuk and @davidrissato --- if this fails, should it be retried?

I am being careful not to worse s3mi's reliability.

**Pull request merged**